srml/grandpa/src/lib.rs

```diff
@@ -0,0 +1,296 @@
+
+#[macro_use]
+extern crate srml_support as runtime_support;
+
+#[cfg(feature = "std")]
+#[macro_use]
+extern crate serde_derive;
+
+extern crate parity_codec;
+#[macro_use]
+extern crate parity_codec_derive;
+
+extern crate sr_primitives as primitives;
+extern crate parity_codec as codec;
+extern crate srml_system as system;
+extern crate srml_session as session;
+extern crate substrate_primitives;
+
+#[cfg(test)]
+extern crate sr_io as runtime_io;
+
+// re-export since this is necessary for `impl_apis` in runtime.
+pub extern crate substrate_finality_grandpa_primitives as fg_primitives;
+
+use rstd::prelude::*;
+use fg_primitives::ScheduledChange;
+use runtime_support::Parameter;
+use runtime_support::dispatch::Result;
+use runtime_support::storage::StorageValue;
+use runtime_support::storage::unhashed::StorageVec;
+use primitives::traits::{
+	MaybeSerializeDebug, CurrentHeight, Convert,
+};
+use substrate_primitives::AuthorityId;
+use system::ensure_signed;
+
+mod mock;
+mod tests;
+
+struct AuthorityStorageVec<S: codec::Codec + Default>(rstd::marker::PhantomData<S>);
+impl<S: codec::Codec + Default> StorageVec for AuthorityStorageVec<S> {
+	type Item = (S, u64);
+	const PREFIX: &'static [u8] = ::fg_primitives::well_known_keys::AUTHORITY_PREFIX;
+}
+
+/// The log type of this crate, projected from module trait type.
+pub type Log<T> = RawLog<
+	<T as system::Trait>::BlockNumber,
+	<T as Trait>::SessionKey,
+>;
+
+/// Logs which can be scanned by GRANDPA for authorities change events.
+pub trait GrandpaChangeSignal<N> {
+	/// Try to cast the log entry as a contained signal.
+	fn as_signal(&self) -> Option<ScheduledChange<N>>;
+}
+
+/// A logs in this module.
+#[cfg_attr(feature = "std", derive(Serialize, Deserialize, Debug))]
+#[derive(Encode, Decode, PartialEq, Eq, Clone)]
+pub enum RawLog<N, SessionKey> {
+	/// Authorities set change has been signalled. Contains the new set of authorities
+	/// and the delay in blocks before applying.
+	AuthoritiesChangeSignal(N, Vec<(SessionKey, u64)>),
+}
+
+impl<N: Clone, SessionKey> RawLog<N, SessionKey> {
+	/// Try to cast the log entry as a contained signal.
+	pub fn as_signal(&self) -> Option<(N, &[(SessionKey, u64)])> {
+		match *self {
+			RawLog::AuthoritiesChangeSignal(ref n, ref signal) => Some((n.clone(), signal)),
+		}
+	}
+}
+
+impl<N, SessionKey> GrandpaChangeSignal<N> for RawLog<N, SessionKey>
+	where N: Clone, SessionKey: Clone + Into<AuthorityId>,
+{
+	fn as_signal(&self) -> Option<ScheduledChange<N>> {
+		RawLog::as_signal(self).map(|(delay, next_authorities)| ScheduledChange {
+			delay,
+			next_authorities: next_authorities.iter()
+				.cloned()
+				.map(|(k, w)| (k.into(), w))
+				.collect(),
+		})
+	}
+}
+
+pub trait Trait: system::Trait {
+	/// Type for all log entries of this module.
+	type Log: From<Log<Self>> + Into<system::DigestItemOf<Self>>;
+
+	/// The session key type used by authorities.
+	type SessionKey: Parameter + Default + MaybeSerializeDebug;
```

Comment:
This should be `MaybeSerializeDebugButNotDeserialize`, I think.

Reply:
We may need deserialization in order to implement `GrandpaApi`, which scans the digest items (decoded)

Reply:
Also doesn't compile, due to an error in `decl_storage!`

Reply:
https://github.com/paritytech/substrate/issues/1148